Additional SecretTypes

jhadvig opened the pull request:
Extended the SecretType with two new types:
- SecretTypeBasicAuth - "kubernetes.io/basic-auth" : 
  - used for basic authentication by providing combination of username and password
- SecretTypeSSHAuth  - "kubernetes.io/ssh-auth" :
  - used for ssh authentication by providing private ssh key

Since the username and password might be empty for some use-cases the validation behaves accordingly.

jhadvig committed:
Additional SecretTypes

mikedanese commented:
@kubernetes/kube-api @kubernetes/rh-platform-management

smarterclayton commented:
ok to test

smarterclayton commented:
@erictune please look as well, we are adding support for these types for git clones and other remote access patterns.  Wanted to settle on official constant names and pattern so folks don't invent different names.

smarterclayton commented:
@erictune whenever you get a chance want your buy in

erictune commented:
LGTM

erictune commented:
A mention of available types added to docs/user-guide/secrets.md would be nice.

thockin commented:
Needs rebase

liggitt commented:
I think the bot has lost its mind. Is it the merge conflict?

mikedanese commented:
@lavalamp @brendandburns

jhadvig commented:
@thockin rebased

jhadvig commented:
@mikedanese thanks :+1:

eparis commented:
From shippable...

```
E0920 17:00:56.136103   10651 util.go:73] Recovered from panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
    /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/util.go:67
    /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/util.go:55
    /usr/local/go/src/runtime/asm_amd64.s:402
    /usr/local/go/src/runtime/panic.go:387
    /usr/local/go/src/runtime/panic.go:42
    /usr/local/go/src/runtime/sigpanic_unix.go:26
    <autogenerated>:58
    /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/cache/listwatch.go:48
 /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/cache/listwatch.go:68
 /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/cache/reflector.go:179
 /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/cache/reflector.go:144
    /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/util.go:116
    /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/util.go:117
    /usr/local/go/src/runtime/asm_amd64.s:2232
```

eparis commented:
Maybe more likely to this PR:
```
    === RUN TestServiceAccountTokenAuthentication

    W0920 17:06:21.703441   16250 controller.go:212] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[]} [{[{172.17.3.217 <nil>}] [] [{ 6443 TCP}]}]}

    W0920 17:06:26.197315   16250 master.go:261] Network range for service cluster IPs is unspecified. Defaulting to 10.0.0.0/24.

    I0920 17:06:26.197638   16250 master.go:287] Node port range unspecified. Defaulting to 30000-32767.

    I0920 17:06:26.202242   16250 master.go:309] Will report 172.17.3.217 as public IP address.

    E0920 17:06:34.613149   16250 handlers.go:37] Unable to authenticate the request due to an error: Token has been invalidated

    E0920 17:06:34.914210   16250 handlers.go:37] Unable to authenticate the request due to an error: Token has been invalidated

    E0920 17:06:35.319816   16250 handlers.go:37] Unable to authenticate the request due to an error: Token has been invalidated

    E0920 17:06:35.741793   16250 handlers.go:37] Unable to authenticate the request due to an error: Token has been invalidated

    panic: test timed out after 2m0s
```

jhadvig commented:
@eparis I don't really see what part of the change could cause the pasted issue

lavalamp commented:
What is the benefit of putting these types in our system explicitly, as opposed to letting people just do whatever they want? (the system doesn't currently prevent them from doing this, IIUC.)

I thought we only were adding these to the list if some part of our system was *genenerating* the relevant secret. Is that the case for these?

(I'm not necessarily opposed to standardizing these, either, I guess, but I'd like some justification/documentation.)

erictune commented:
We also standardize formats that the system will _consume_, such as imagePullSecrets.  Since "system" is a broad term, and might encompass "plugins", then I think we need to define a bunch of common formats.

smarterclayton commented:
We were planning on adding generators / behaviors for a couple of these use
cases in OpenShift.  We would like to get people thinking about using a
standard pattern for these specifically so that we don't do ad hoc.
Practically, I expect these to be used commonly as simple descriptive
formats.

On Mon, Sep 21, 2015 at 1:58 PM, Eric Tune <notifications@github.com> wrote:

> We also standardize formats that the system will *consume*, such as
> imagePullSecrets. Since "system" is a broad term, and might encompass
> "plugins", then I think we need to define a bunch of common formats.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13853#issuecomment-142061354>
> .
>

smarterclayton commented:
Described that here https://github.com/kubernetes/kubernetes/pull/13853#issuecomment-139683248

bgrant0607 commented:
Got tests?

jhadvig commented:
@bgrant0607 tests delivered. PTAL

jhadvig commented:
@bgrant0607 @mikedanese can you please look again ? Thanks

smarterclayton commented:
LGTM although @bgrant0607 needs to sign off on the API change.

erictune commented:
The integration test failure is a flake.  

@bgrant0607 do you want to comment on the addition of some string constants to the API?

bgrant0607 commented on pkg/api/types.go:
```diff
@@ -1882,6 +1882,27 @@ const (
+	SecretTypeSSHAuth SecretType = "kubernetes.io/ssh-auth"
+
+	// SSHAuthPrivateKey is the key of the required SSH private key for SecretTypeSSHAuth secrets
+	SSHAuthPrivateKey = "ssh-privatekey"
```
Is there a reason why "ssh-privatekey" is not "ssh-private-key"?

erictune replied:
@jhadvig any reason?

erictune replied:
Rename this or explain why funny hyphenation.

jhadvig replied:
As @smarterclayton already explained:
For ssh-privatekey, "privatekey" is a common identifier used by SSH config, that's the only reason.

smarterclayton replied:
If we have to pick an arbitrary value, and we have a pattern for key types,
it was primarily about picking something that clearly indicated what the
type was that fit into that (define a convention consistent with existing
conventions).

On Sat, Dec 19, 2015 at 7:22 AM, Jakub Hadvig <notifications@github.com>
wrote:

> In pkg/api/types.go
> <https://github.com/kubernetes/kubernetes/pull/13853#discussion_r48091879>
> :
>
> > +	// - Secret.Data["password"] - password or token needed for authentication
> > +	SecretTypeBasicAuth SecretType = "kubernetes.io/basic-auth"
> > +
> > +	// BasicAuthUsernameKey is the key of the username for SecretTypeBasicAuth secrets
> > +	BasicAuthUsernameKey = "username"
> > +	// BasicAuthPasswordKey is the key of the password or token for SecretTypeBasicAuth secrets
> > +	BasicAuthPasswordKey = "password"
> > +
> > +	// SecretTypeSSHAuth contains data needed for SSH authetication.
> > +	//
> > +	// Required field:
> > +	// - Secret.Data["ssh-privatekey"] - private SSH key needed for authentication
> > +	SecretTypeSSHAuth SecretType = "kubernetes.io/ssh-auth"
> > +
> > +	// SSHAuthPrivateKey is the key of the required SSH private key for SecretTypeSSHAuth secrets
> > +	SSHAuthPrivateKey = "ssh-privatekey"
>
> As @smarterclayton <https://github.com/smarterclayton> already explained:
> For ssh-privatekey, "privatekey" is a common identifier used by SSH
> config, that's the only reason.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13853/files#r48091879>.
>

bgrant0607 commented:
Changes also need to be made in api/v1/types.go, and need to be documented in docs/user-guide/secrets.md.

bgrant0607 commented on pkg/api/validation/validation.go:
```diff
@@ -1603,6 +1603,21 @@ func ValidateSecret(secret *api.Secret) errs.ValidationErrorList {
 		if err := json.Unmarshal(dockercfgBytes, &map[string]interface{}{}); err != nil {
 			allErrs = append(allErrs, errs.NewFieldInvalid(fmt.Sprintf("data[%s]", api.DockerConfigKey), "<secret contents redacted>", err.Error()))
 		}
+	case api.SecretTypeBasicAuth:
```
If Kubernetes itself is not consuming these secret types in any way, then we shouldn't validate them in the apiserver.

smarterclayton replied:
For ssh-privatekey, "privatekey" is a common identifier used by SSH config,
that's the only reason.

On Mon, Oct 19, 2015 at 3:58 PM, Brian Grant <notifications@github.com>
wrote:

> In pkg/api/validation/validation.go
> <https://github.com/kubernetes/kubernetes/pull/13853#discussion_r42417980>
> :
>
> > @@ -1603,6 +1603,21 @@ func ValidateSecret(secret *api.Secret) errs.ValidationErrorList {
> >  		if err := json.Unmarshal(dockercfgBytes, &map[string]interface{}{}); err != nil {
> >  			allErrs = append(allErrs, errs.NewFieldInvalid(fmt.Sprintf("data[%s]", api.DockerConfigKey), "<secret contents redacted>", err.Error()))
> >  		}
> > +	case api.SecretTypeBasicAuth:
>
> If Kubernetes itself is not consuming these secret types in any way, then
> we shouldn't validate them in the apiserver.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13853/files#r42417980>.
>

bgrant0607 commented:
Basic auth and ssh are common enough that I'm fine with adding special cases for them. 

However, my previous comments still stand:

* need to be added to v1/types.go also
* needs documentation in the user guide

Please ensure the validation errors are clear enough for the validation to be worthwhile.

jhadvig commented:
@bgrant0607 will try to deliver those asap

jhadvig commented:
@bgrant0607 can you please point me out to the secret docs that shall be updated. I found several of them and not sure on which places the update shall be done. Thanks !

smarterclayton commented:
I though we said we are keeping constants only in the core types file?  Is
that a change from previous?

On Nov 30, 2015, at 1:25 PM, Brian Grant <notifications@github.com> wrote:

Basic auth and ssh are common enough that I'm fine with adding special
cases for them.

However, my previous comments still stand:

   - need to be added to v1/types.go also
   - needs documentation in the user guide

Please ensure the validation errors are clear enough for the validation to
be worthwhile.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13853#issuecomment-160714513>
.

bgrant0607 commented:
@smarterclayton We're totally inconsistent about which constants are in v1/types.go and which aren't. Secret types currently are, so all should be or none should be.

bgrant0607 commented:
docs/user-guide/secrets.md

jhadvig commented:
@bgrant0607 I've added the new secret types to v1/types and updated the docs with list of the secret for users along with description and requirements. Wasn't sure if `SecretTypeServiceAccountToken` shall be listed as well, so I've skipped it. 

Also I've noticed that `SecretTypeDockerConfigJson` secret type is only present in api/types.go and not in api/v1/types.go. Is that intentional ?

erictune commented:
@deads2k might know differently, but I think all secret types belong in both files.

erictune commented:
Thanks for adding documentation.  There is one unanswered question.   Also it would be good to copy to the other type.go in this or a subsequent PR.  Very close.

jhadvig commented:
@erictune its your call. I can do it in this one, you just need to tell me , in which one would you like to see those types.
Thanks !

smarterclayton commented:
@jhadvig can you get this closed out? :)

erictune commented:
LGTM

erictune commented:
Build failed:
```
# k8s.io/kubernetes/pkg/api/validation
_output/local/go/src/k8s.io/kubernetes/pkg/api/validation/validation.go:1907: undefined: validation.NewRequiredError
_output/local/go/src/k8s.io/kubernetes/pkg/api/validation/validation.go:1908: undefined: validation.NewRequiredError
_output/local/go/src/k8s.io/kubernetes/pkg/api/validation/validation.go:1913: undefined: validation.NewRequiredError
!!! Error in /home/travis/gopath/src/github.com/kubernetes/kubernetes/hack/lib/golang.sh:395
  'go install "${goflags[@]:+${goflags[@]}}" -ldflags "${goldflags}" "${nonstatics[@]:+${nonstatics[@]}}"' exited with status 2
```

erictune commented:
Does this need to also change `kubectl` to support `kubectl create secret basic-auth` and `kubectl create secret `ssh-auth`?

jhadvig commented:
@erictune I've updated the PR.
In order to support the two new secret types, I think that the `kubectl` has to be updated. But would rather do it as a follow up.
Thoughts ?

erictune commented:
Yes, doing in a follow up PR is fine.
